### PR TITLE
feat: kudoctl configuration via file and environment

### PR DESCRIPTION
**Pull request by NilsPonsard**
This PR adds a `Config` struct used to store the configuration variables of the program.
A function that reads the file and the environment is also provided.

### CODE REVIEW

File: kudoctl/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        if let Some(parent) = parent {
+            create_dir_all(parent)?;
+        }
+
+        let config_file = File::create(path)?;
+
+        let default = ConfigFile {
+            controller_url: default_controller_url(),
+            verbosity_level: default_log_level_str(),
+        };
```
Comment (sameo) [inline kudoctl/src/config.rs:68]: You want to implement the `Default` trait for `ConfigFile`.

File: kudoctl/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+pub struct Config {
+    config_file: PathBuf,
+    pub controller_url: String,
+    pub verbosity_level: LevelFilter,
+}
+
+// Read the config file and return a Config object.
+// If the file does not exist, creates one with the default values.
+fn read_config_file(path: &PathBuf) -> Result<ConfigFile, Box<dyn std::error::Error>> {
+    if !path.exists() {
```
Comment (sameo) [inline kudoctl/src/config.rs:56]: The following construct is more idiomatic:

```rust
let config_file = if !path.exists() {
    ...
    Default::default()
} else {
     let file = File::open(path)?;
     serde_yaml::from_reader(file)?
};
```

File: kudoctl/src/config.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+// Read the configuration from the config file and the environment variables.
+// The environment variables override the values in the config file.
+pub fn read_config(file: String) -> Result<Config, Box<dyn std::error::Error>> {
+    // Read the config file
+
+    let file_path = if let Ok(path) = env::var("KUDO_CONFIG") {
+        Path::new(&path).to_path_buf()
+    } else {
+        default_config_file()
+    };
```
Comment (sameo) [inline kudoctl/src/config.rs:88]: Use `unwrap_or` or `unwrap_or_else` here.

File: kudoctl/src/config.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        verbosity_level,
+    })
+}
+
+// Reads the environment variable and returns the value if it is set, returns the `config_var` otherwise.
+fn check_env_override(env_var: &str, config_var: &str) -> String {
+    if let Ok(env_var) = env::var(env_var) {
+        env_var
+    } else {
+        config_var.to_string()
```
Comment (sameo) [inline kudoctl/src/config.rs:114]: Same here: Use `unwrap_or_else`

File: kudoctl/src/config.rs (hunk 2 above)
Comment (NilsPonsard) [reply kudoctl/src/config.rs]: I tried to make it more idiomatic like you asked, but I'm not sure if the way the default config is written is good enough.

**Comment by sameo** [comment]
You're missing a SOB in your last commit. Please squash it into the relevant commit from this PR instead.

**Comment by sameo** [comment]
@NilsPonsard Please squash your 2 last `refactor` commits into the relevant previous one, and we'll merge that PR.

**Comment by NilsPonsard** [comment]
Squashed

**Review by sameo — APPROVED** [review]

**Pull request merged**